src/writer.rs

```diff
@@ -0,0 +1,450 @@
+            if root_bytes <= target_root_len {
+                return Ok((root_dir, 0));
+            }
+        }
+
+        // TODO: case 2: mixed tile entries/directory entries in root
+
+        // case 3: root directory is leaf pointers only
+        // use an iterative method, increasing the size of the leaf directory until the root fits
+
+        let mut leaf_size = (self.entries.len() / 3500).max(4096);
+        loop {
+            let (root_dir, num_leaves) = self.build_roots_leaves(leaf_size)?;
+            let root_bytes = root_dir.compressed_size(self.header.internal_compression)?;
+            if root_bytes <= target_root_len {
+                return Ok((root_dir, num_leaves));
+            }
+            leaf_size += leaf_size / 5; // go-pmtiles: leaf_size *= 1.2
+        }
+    }
+    fn build_roots_leaves(&self, leaf_size: usize) -> PmtResult<(Directory, usize)> {
+        let mut root_dir = Directory::with_capacity(self.entries.len() / leaf_size);
+        let mut offset = 0;
+        for chunk in self.entries.chunks(leaf_size) {
+            let leaf_size = self.dir_size(chunk)?;
+            root_dir.push(DirEntry {
+                tile_id: chunk[0].tile_id,
+                offset,
+                length: into_u32(leaf_size)?,
+                run_length: 0,
+            });
+            offset += leaf_size as u64;
+        }
+
+        let num_leaves = root_dir.entries().len();
+        Ok((root_dir, num_leaves))
+    }
+    fn dir_size(&self, entries: &[DirEntry]) -> PmtResult<usize> {
+        let dir = Directory::from_entries(entries);
+        dir.compressed_size(self.header.internal_compression)
+    }
+
+    /// Finish writing the `PMTiles` file.
+    pub fn finish(mut self) -> PmtResult<()> {
+        if let Some(last) = self.entries.last() {
+            self.header.data_length = last.offset + u64::from(last.length);
+            self.header.leaf_offset = self.header.data_offset + self.header.data_length;
+            self.header.n_addressed_tiles = self.n_addressed_tiles.try_into().ok();
+            self.header.n_tile_entries = (self.entries.len() as u64).try_into().ok();
+            self.header.n_tile_contents = self.n_tile_contents.try_into().ok();
+        }
+        // Write leaf directories and get root directory
+        let root_dir = self.build_directories()?;
+        // Determine compressed root directory length
+        let mut root_dir_buf = vec![];
+        root_dir.write_compressed_to(&mut root_dir_buf, self.header.internal_compression)?;
+        self.header.root_length = root_dir_buf.len() as u64;
+
+        // Write header and root directory
+        self.out.rewind()?;
+        self.header.write_to(&mut self.out)?;
+        self.out.write_all(&root_dir_buf)?;
+        self.out.flush()?;
+
+        Ok(())
+    }
+}
+
+fn compare_blobs(a: &[u8], b: &[u8]) -> bool {
+    a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| a == b)
+}
+
+fn into_u32(v: usize) -> PmtResult<u32> {
```

Comment:
it seems this is only used twice - should it be a func?  I'm ok either way

Reply:
Maybe it also useful for the reader, where currently `allow(clippy::cast_possible_truncation)` is set.